Documentation of p2p layer in Tendermint v0.34 (#9348)
tendermint/tendermint: spec/p2p/v0.34/addressbook.md

# Address Book

The address book tracks information about other nodes in the network.

The primary information about a node is its peer address,

> composed by a network address and an ID.

Reviewer: ```suggestion
composed of a network address and an ID.
```

Addresses can be manually configured (e.g., as persistent peers),
can be learned when a peer attempts to connect to the node,
or can be retrieved via the [Peer Exchange protocol](./pex-protocol.md).

The [Peer Manager](./peer_manager.md) retrieves from the address book peer
addresses to dial, so to establish outbound connections.
It also records in the address book failed attempts to connect to a peer.

> The [Peer Exchange protocol](./pex-protocol.md) feds the address book with

Reviewer: ```suggestion
The [Peer Exchange protocol](./pex-protocol.md) feeds the address book with
```

addresses received from peers, and retrieves from the address book random
selection of addresses it provides to peers.

Once the node is connected to a peer, it uses the address book to record
information about the quality of the peer.
The quality of a peer is currently restricted to the definition of
[good](#good-peers) or [bad](#bad-peers) peers.
It influences, in particular, the selection of candidate peers to dial.

## New addresses

The `AddAddress` method adds a new address to the address book.

> The added address is associated to a *source* address, which identifies the

Reviewer: Is source address the address of the peer from which new address is learned (not necessarily that peer address)? For example if I learned from peer A address of peer B, then source address for B would be A. right?

node from which the address was learned.

Addresses are added to the address book in the following situations:

1. When a peer address is received via PEX protocol, the source is the sender
   of the PEX message
1. When an inbound peer is added to the PEX reactor, in this case the peer
   itself is set as the source of its address
1. When the switch is instructed to dial addresses via the `DialPeersAsync`
   method, in this case the node itself is set as the source

An address contains a node ID and a network address (IP and port).
The address book can keep multiple network addresses for the same node ID, but
this number should be limited.
A new address with a node ID that is already present in the address book is

> **not added** when:

Reviewer: This is an example where higher level invariant should be captured, i.e., intention behind this mechanism, and not just rules, as it is not really clear why this mechanism exists and what it is trying to achieve.

> - The last address added with the same node ID is stored in an old bucket, so

Reviewer: Bucket is not defined here so it is hard to follow this paragraph.

  it is considered a "good" address
- There are addresses associated to the same node ID stored in
  `maxNewBucketsPerAddress = 4` distinct buckets
- Randomly, with a probability that increases exponentially with the number of
  buckets in which there is an address with the same node ID.
  So, a new address for a node ID which is already present in one bucket is
  added with 50% of probability; if the node ID is present in two buckets, the
  probability decreases to 25%; and if it is present in three buckets, the
  probability is 12.5%.

Addresses are stored in [buckets](#buckets).
The bucket to which a new address is added is computed based on the address
itself and its source address.
An address is not added to a bucket if it is already present there.
This should not happen with addresses with new node IDs, but it can happen with
addresses with known node IDs.
In fact, as the mapping of buckets for new addresses is mainly determined by
the network group of the source address, the same address can be present in
more than one bucket provided that it was added with different source addresses.

> The new address is also added to the `addrLookup` table, which stores

Reviewer: This is too low level and in my view not really very helpful. We should try to capture spec of the address book at the more high level: what is the purpose of this structure, what are the main APIs and what invariants it should maintain? The pseudocode like description should in my view not be part of the spec but lives as a comment in the code and its purpose should be to help reader understands the code. Spec should defines higher level concepts and language we used when we talk about main p2p features.

`knownAddress` entries indexed by their node IDs.
If the new address is from an unknown peer, a new entry is added to the
`addrLookup` table; otherwise, the existing entry is updated with the new
address.
Entries of this table contain, among other fields, the list of buckets where
addresses of a peer are stored.
The `addrLookup` table is used by most of the address book methods (e.g.,
`HasAddress`, `IsGood`, `MarkGood`, `MarkAttempt`), as it provides fast access
to addresses.

> ### Errors

Reviewer 2: I think this is a mistake, it should be either the new address is the address of the node adding the address. @cason , am I correct?

- if `routabilityStrict` is set and the address is not routable
- in case of failures computing the bucket for the new address (`calcNewBucket` method)
- if the added address instance, which is a new address, is configured as an
  old address (sanity check of `addToNewBucket` method)

## Need for Addresses

The `NeedMoreAddrs` method verifies whether the address book needs more addresses.

It is invoked by the PEX reactor to define whether to request peer addresses
to a new outbound peer or to a randomly selected connected peer.

The address book needs more addresses when it has less than `1000` addresses
registered, counting all buckets for new and old addresses.

## Pick address

The `PickAddress` method returns an address stored in the address book, chosen
at random with a configurable bias toward new addresses.

It is invoked by the PEX reactor to obtain a peer address to dial, as part of
its `ensurePeers` routine.
The bias starts from 10%, when the peer has no outbound peers, increasing by
10% for each outbound peer the node has, up to 90%, when the node has at least
8 outbound peers.

> The configured bias is a parameter that influences the probability of choosing

Reviewer: The same comment as above; too low level and not clear what  this logic is trying to achieve.

It is invoked by the consensus reactor, via switch, when the number of useful
messages received from a peer is a multiple of `10000`.
Vote and block part messages are considered for this number, they must be valid
and not be duplicated messages to be considered useful.

> The `SwitchReporter` type of `behaviour` package also invokes the `MarkGood`
> method when a "reason" associated with consensus votes and block parts is
> reported.
> No reactor, however, currently provides these "reasons" to the `SwitchReporter`.

The effect of this action is that the address registered for the peer's ID in the
`addrLookup` table, which is the last address added with that ID, is marked as
good and moved to a bucket of old addresses.
An address marked as good has its failed to connect counter and timestamp reset.
If the destination bucket of old addresses is full, the oldest address in the
bucket is moved (downgraded) to a bucket of new addresses.

Moving the peer address to a bucket of old addresses has the effect of
upgrading, or increasing the ranking of a peer in the address book.

## Bad peers

The `MarkBad` method marks a peer as bad and bans it for a period of time.

> 

Reviewer 2: ```suggestion

This method is invoked from two places: within the [dial routine](#dial-attempts) or from the PEX reactor. 

In case of dialing peers, a peer is marked as bad if there are authentication errors. Other types of errors cause a simple disconnect from the peer. 

```

The [Peer Manager](./peer_manager.md) retrieves addresses to dial, so to
establish outbound connections.
And the [Peer Exchange protocol](./pex-protocol.md) retrieves random samples of
addresses to offer (send) to peers.

In addition to addresses, the address book also stores some information about
the quality of peers with which the node has interacted.
This includes information about connection attempts, failed or succeeded, and
quality information provided by the several protocols running on the node.
Currently, the quality of a peer is restricted to three states: new,
[good](#good-peers), or [bad](#bad-peers).
The quality metrics associated to peers influences the selection of peers to
dial and offer via PEX protocol, in a way that they are not fully random.

## Adding addresses

The `AddAddress` method adds the address of a peer to the address book.

The added address is associated to a *source* address, which identifies the
node from which the peer address was learned.

Addresses are added to the address book in the following situations:

1. When a peer address is learned via PEX protocol, having the sender
   of the PEX message as its source

> 1. When an inbound peer is added, in this case the peer itself is set as the

Reviewer 2: ```suggestion
2. When an inbound peer is added, in this case the peer itself is set as the
```

   source of its own address

> 1. When the switch is instructed to dial addresses via the `DialPeersAsync`

Reviewer 2: ```suggestion
3. When the switch is instructed to dial addresses via the `DialPeersAsync`
```

   method, in this case the node itself is set as the source

Addresses are stored in [buckets](#buckets).
There are buckets for new addresses and buckets for old addresses.
If the added address belongs to a new peer, i.e., to a node ID that is not
present in the address book, it is added to a bucket of new addresses.

The rationale of this organization of addresses into buckets is detailed

> thorough this document, but it can summarized as follows.

Reviewer 2: ```suggestion
thorough this document, but it can be summarized as follows.
```

A new peer has its address stored in a bucket of new address, as the node does

> not has much information about it.

Reviewer 2: ```suggestion
not have much information about it.
```

When a peer is marked as a [good peer](#good-peers) by some reactor (protocol),
its address is moved (upgraded) to a bucket of old addresses.
Old addresses are therefore of peers about which the node has good information.
When a bucket of old addresses becomes full, one of its addresses is moved
(downgraded) to a bucket of new addresses.

> When a bucket of new address becomes full, one of its address is dropped.

Reviewer 2: ```suggestion
When a bucket of new addresses becomes full, one of its addresses is dropped.
```

This should not happen with addresses with new node IDs, but it can happen with
addresses with known node IDs.
In fact, as the mapping of buckets for new addresses is mainly determined by
the network group of the source address, the same address can be present in
more than one bucket provided that it was added with different source addresses.

The new address is also added to the `addrLookup` table, which stores
`knownAddress` entries indexed by their node IDs.
If the new address is from an unknown peer, a new entry is added to the
`addrLookup` table; otherwise, the existing entry is updated with the new
address.
Entries of this table contain, among other fields, the list of buckets where
addresses of a peer are stored.
The `addrLookup` table is used by most of the address book methods (e.g.,
`HasAddress`, `IsGood`, `MarkGood`, `MarkAttempt`), as it provides fast access
to addresses.

> TODO: reorganize this description having the goals of this organization
> explained before their actual implementation.

### Errors

- if the added address or the associated source address are nil
- if the added address is invalid
- if the added address is the local node's address

> - if the added address ID is of a banned peer

Reviewer 3: "banned" is not defined at this point. Please add a forward reference to the "bad peers" section

- if either the added address or the associated source address IDs are configured as private IDs
- if `routabilityStrict` is set and the address is not routable
- in case of failures computing the bucket for the new address (`calcNewBucket` method)
- if the added address instance, which is a new address, is configured as an
  old address (sanity check of `addToNewBucket` method)

## Need for Addresses

The `NeedMoreAddrs` method verifies whether the address book needs more addresses.

It is invoked by the PEX reactor to define whether to request peer addresses
to a new outbound peer or to a randomly selected connected peer.

The address book needs more addresses when it has less than `1000` addresses
registered, counting all buckets for new and old addresses.

## Pick address

The `PickAddress` method returns an address stored in the address book, chosen

> at random with a configurable bias toward new addresses.

Reviewer 2: ```suggestion
at random with a configurable bias towards new addresses.
```

As a result, it becomes more likely to pick a new address at random from this sample.

> The use of the square roots softens the impact of disproportional numbers of
> new and old addresses in the address book. This is actually the expected
> scenario, as there are 4 times more buckets for new addresses than buckets
> for old addresses.

Once the type of address, new or old, is defined, a non-empty bucket of this
type is selected at random.
From the selected bucket, an address is chosen at random and returned.
If all buckets of the selected type are empty, no address is returned.

## Random selection

The `GetSelection` method returns a selection of addresses stored in the
address book, with no bias toward new or old addresses.

It is invoked by the PEX protocol to obtain a list of peer addresses with two
purposes:

- To send to a peer in a PEX response, in the case of outbound peers or of
  nodes not operating in seed mode
- To crawl, in the case of nodes operating in seed mode, as part of every
  interaction of the `crawlPeersRoutine`

The selection is a random subset of the peer addresses stored in the

> `addrLookup` table, which stores that last address added for each peer ID.

Reviewer 2: ```suggestion
`addrLookup` table, which stores the last address added for each peer ID.
```

The target size of the selection is `23%` (`getSelectionPercent`) of the
number of addresses stored in the address book, but it should not be lower than
`32` (`minGetSelection`) --- if it is, all addresses in the book are returned
--- nor greater than `250` (`maxGetSelection`).

> The random selection is produced by:
> - Retrieving all entries of the `addrLookup` map, which by definition are
>   returned in random order.
> - Randomly shuffling the retrieved list, using the Fisher-Yates algorithm

## Random selection with bias

The `GetSelectionWithBias` method returns a selection of addresses stored in
the address book, with bias toward new addresses.

It is invoked by the PEX protocol to obtain a list of peer addresses to be sent
to a peer in a PEX response.
This method is only invoked by seed nodes, when replying to a PEX request
received from an inbound peer (i.e., a peer that dialed the seed node).
The bias used in this scenario is hard-coded to 30%, meaning that 70% of
the returned addresses are expected to be old addresses.

The number of addresses that compose the selection is computed in the same way

> as for then non-biased random selection.

Reviewer 2: ```suggestion
as for the non-biased random selection.
```

This list of all addresses of a type is randomly shuffled, and the requested
number of addresses are retrieved from the tail of this list.
The returned selection contains, at its beginning, a random selection of new
addresses in random order, followed by a random selection of old addresses, in
random order.

## Dial Attempts

The `MarkAttempt` method records a failed attempt to connect to an address. 

It is invoked by the Peer Manager when it fails dialing a peer, but the failure
is not in the authentication step (`ErrSwitchAuthenticationFailure` error).
In case of authentication errors, the peer is instead marked as a [bad peer](#bad-peers).

The failed connection attempt is recorded in the address registered for the
peer's ID in the `addrLookup` table, which is the last address added with that ID.
The known address' counter of failed `Attempts` is increased and the failure
time is registered in `LastAttempt`.

The possible effect of recording multiple failed connect attempts to a peer is
to turn its address into a *bad* address (do not confuse with banned addresses).
A known address becomes bad if it is stored in buckets of new addresses, and

> when connections attempts:

Reviewer 2: ```suggestion
when connection attempts:
```

>
> More precisely, failed connection attempts are recorded in the entry of  the
> `addrLookup` table with reported peer ID, which contains the last address
> added for that node ID, which is not necessarily the reported peer address.

## Good peers

The `MarkGood` method marks a peer ID as good.

It is invoked by the consensus reactor, via switch, when the number of useful
messages received from a peer is a multiple of `10000`.
Vote and block part messages are considered for this number, they must be valid
and not be duplicated messages to be considered useful.

> The `SwitchReporter` type of `behaviour` package also invokes the `MarkGood`
> method when a "reason" associated with consensus votes and block parts is
> reported.
> No reactor, however, currently provides these "reasons" to the `SwitchReporter`.

The effect of this action is that the address registered for the peer's ID in the
`addrLookup` table, which is the last address added with that ID, is marked as
good and moved to a bucket of old addresses.
An address marked as good has its failed to connect counter and timestamp reset.
If the destination bucket of old addresses is full, the oldest address in the

> bucket is moved (downgraded) to a bucket of new addresses.

Reviewer 3: Can't this cause a periodic churn between buckets if all "good" peers don't fit their corresponding "old" buckets? This churn would be anything but useful, BTW.

Moving the peer address to a bucket of old addresses has the effect of
upgrading, or increasing the ranking of a peer in the address book.

> **Note** In v0.34 a peers is currently marked good only from the consensus reactor 

Reviewer 3: ```suggestion
**Note** In v0.34 a peer is currently marked good only from the consensus reactor 
```

whenever it delivers a correct consensus message.

## Bad peers

The `MarkBad` method marks a peer as bad and bans it for a period of time.

It is invoked by the [PEX reactor](pex-protocol.md#misbehavior), with banning time of 24 hours, in the following cases:

- When PEX requests are received too often from a peer
- When an invalid PEX response is received from a peer
- When an unsolicited PEX response is received from a peer

> - When the `maxAttemptsToDial` a limit (`16`) is reached for a peer

Reviewer 3: ```suggestion
- When the `maxAttemptsToDial` limit (`16`) is reached for a peer
```

- If an `ErrSwitchAuthenticationFailure` error is returned when dialing a peer

The effect of this action is that the address registered for the peer's ID in the
`addrLookup` table, which is the last address added with that ID, is banned for
a period of time.
The banned peer is removed from the `addrLookup` table and from all buckets
where its addresses are stored.

The information about banned peers, however, is not discarded.
It is maintained in the `badPeers` map, indexed by peer ID.
This allows, in particular, addresses of banned peers to be
[reinstated](#reinstating-addresses), i.e., to be added
back to the address book, when their ban period expires.

## Reinstating addresses

The `ReinstateBadPeers` method attempts to re-add banned addresses to the address book.

It is invoked by the PEX reactor when dialing new peers.
This action is taken before requesting additional addresses to peers,
in the case that the node needs more peer addresses.

The set of banned peer addresses is retrieved from the `badPeers` map.

> Addresses that are not any longer banned, i.e., which banned period has expired,

Reviewer 2: ```suggestion
Addresses that are not any longer banned, i.e., whose banned period has expired,
```

are added back to the address book as new addresses, while the corresponding
node IDs are removed from the `badPeers` map.

> ## Removing addresses

Reviewer 3: Is there a mechanism for a _really_ bad address (simple example: an address that didn't _ever_ work for, say, months) to be deleted from the address book altogether?
I'm thinking of a malicious actor (e.g. a malicious seed node) providing plenty of _fake_ addresses that will then be kept around forever

Reviewer 3: Ok, I see there is such mechanism in the `Buckets` section. I guess I agree with @adizere in that (some form of) the `Buckets` section should appear early on in this file.

The `RemoveAddress` method removes an address from the address book.

It is invoked by the switch when it dials a peer or accepts a connection from a

> peer that end ups being the node itself (`IsSelf` error).

Reviewer 3: ```suggestion
peer that ends up being the node itself (`IsSelf` error).
```

The [Peer Exchange protocol](./pex-protocol.md) stores in the address book
the peer addresses it discovers, i.e., it learns from connected peers.
And the [Switch](./switch.md) registers the addresses of peers with which it
has interacted: to which it has dialed or from which it has accepted a
connection.

The address book also records additional information about peers with which the
node has interacted, from which is possible to rank peers.
The Switch reports [connection attempts](#dial-attempts) to a peer address; too
much failed attempts indicate that a peer address is invalid.
Reactors, in they turn, report a peer as [good](#good-peers) when it behaves as
expected, or as a [bad peer](#bad-peers), when it misbehaves.

There are two entities that retrieve peer addresses from the address book.
The [Peer Manager](./peer_manager.md) retrieves peer addresses to dial, so to
establish outbound connections.
This selection is random, but has a configurable bias towards peers that have
been marked as good peers.
The [Peer Exchange protocol](./pex-protocol.md) retrieves random samples of
addresses to offer (send) to peers.

> This selection is also random but it include, in particular for nodes that

Reviewer 2: ```suggestion
This selection is also random but it includse, in particular for nodes that
```

operate in seed mode, some bias toward peers marked as good ones.

## Buckets

> Peer addresses are stored into buckets.

Reviewer 2: ```suggestion
Peer addresses are stored in buckets.
```

the bucket is removed from the address book.
In other words, exceeding old or good addresses are downgraded to new
addresses, while exceeding new addresses are dropped.

The bucket that stores an `address` is defined by the following two methods,
for new and old addresses:

- `calcNewBucket(address, source) = hash(key + groupKey(source) + hash(key + groupKey(address) + groupKey(source)) % newBucketsPerGroup) % newBucketCount`
- `calcOldBucket(address) = hash(key + groupKey(address) + hash(key + address) % oldBucketsPerGroup) % oldBucketCount`

The `key` is a fixed random 96-bit (8-byte) string.
The `groupKey` for an address is a string representing its network group.
The `source` of an address is the address of the peer from which we learn the
address..
The first (internal) hash is reduced to an integer up to `newBucketsPerGroup =
32`, for new addresses, and `oldBucketsPerGroup = 4`, for old addresses.
The second (external) hash is reduced to bucket indexes, in the interval from 0
to the number of new (`newBucketCount = 256`) or old (`oldBucketCount = 64`) buckets.

Notice that new addresses with sources from the same network group are more

> likely to end up in, therefore to compete for, the same bucket.

Reviewer 2: ```suggestion
likely to end up in the same bucket, therefore to competing for it.
```

been marked as good peers.
The [Peer Exchange protocol](./pex-protocol.md) retrieves random samples of
addresses to offer (send) to peers.

> This selection is also random but it includse, in particular for nodes that

Reviewer 2: ```suggestion
This selection is also random but it includes, in particular for nodes that
```
